janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphPropertiesStep.java

```diff
@@ -82,55 +95,28 @@ public void setUseMultiQuery(boolean useMultiQuery) {
         return (Iterator<E>) Iterators.transform(iterable.iterator(), Property::value);
     }
 
-    private void initialize() {
-        assert !initialized;
-        initialized = true;
-        assert getReturnType().forProperties() || (orders.isEmpty() && hasContainers.isEmpty());
-
-        if (!starts.hasNext()) throw FastNoSuchElementException.instance();
-        final List<Traverser.Admin<Element>> elements = new ArrayList<>();
-        starts.forEachRemaining(elements::add);
-        starts.add(elements.iterator());
-        assert elements.size() > 0;
-
-        useMultiQuery = useMultiQuery && elements.stream().allMatch(e -> e.get() instanceof Vertex);
-
-        if (useMultiQuery) {
-            initializeMultiQuery(elements);
-        }
-    }
-
     /**
-     * This initialisation method is called the first time this instance is used and also when
-     * an attempt to retrieve a vertex from the cached multiQuery results doesn't find an entry.
-     * @param vertices A list of vertices with which to initialise the multiQuery
+     * This initialisation method is called when an attempt to retrieve a vertex from the cached multiQuery results
+     * doesn't find an entry.
      */
-    private void initializeMultiQuery(final List<Traverser.Admin<Element>> vertices) {
-        assert vertices.size() > 0;
-        final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
-        vertices.forEach(v -> multiQuery.addVertex((Vertex)v.get()));
+    private void prefetchNextBatch() {
+        final JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(getTraversal()).multiQuery();
+        multiQuery.addAllVertices(verticesToPrefetch);
+        verticesToPrefetch.clear();
         makeQuery(multiQuery);
 
-        Map<JanusGraphVertex, Iterable<? extends JanusGraphProperty>> results = multiQuery.properties();
-        if (multiQueryResults == null) {
-            multiQueryResults = results;
-        } else {
-            multiQueryResults.putAll(results);
+        try{
+            multiQueryResults = multiQuery.properties();
+        } catch (JanusGraphException janusGraphException) {
```

Comment:
I explicitly added that `try ... catch` because there is some test which asserts that a `TraversalInterruptedException` is thrown on certain occasions.

Reply:
Can you share which test fails? In that case why the old code does not suffer this problem?
Besides, I think we should at least do
```java
if (Thread.interrupted()) throw new TraversalInterruptedException()
```
otherwise it does not seem to be semantically correct

Reply:
This is the test that fails if no `TraversalInterruptedException` is thrown:
https://github.com/apache/tinkerpop/blob/6bb5c19e5b3845a1547f079151017861b47e0828/gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/TraversalInterruptionTest.java#L121

It creates a `sideEffect` which lets the first traverser pass and lets each of the succeeding traversers wait 3s before letting them pass. The test then lets the traversal run for five seconds before interrupting the thread which iterates the traversal. The expectation is that the `traversal.iterate()` call resulted in a `TraversalInterruptedException`.

I think what happened before this PR is that the `JanusGraphVertexStep` was still trying to aggregate its `starts` when the thread was interrupted, but it had not yet started to execute backend queries. Now with this PR, the traversers have already passed the `JanusGraphMultiQueryStep` before entering the `sideEffectStep` and thus the `JanusGraphVertexStep` already knows the set of vertices it needs to prefetch. The exception then somehow occurs during the execution of the MultiQuery.

Reply:
Got it. Now it looks good to me. I am just confused why CodeCov complains line 123 is not covered.

Reply:
But that did not solve the issue. The exception is not thrown in this case, because the current thread is not interrupted. It seems to be some other thread which gets interrupted and thereby raises the `JanusGraphException` during the MultiQuery execution.

Reply:
I found a better solution. In the cases we discussed here, the thrown `JanusGraphException` is caused by an `InterruptedException`. We can simply check `janusGraphException.isCausedBy(InterruptedException.class)`.

Reply:
Somehow Codecov is still reporting that line 123 is not covered by tests. If you set a debug point on line 123 and run the test locally, do you see it get executed?

Reply:
Yes, I did exactly that when I was debugging the exception handling and trying to figure out which Exception has to be caught/thrown. I don't see why Codecov complains here, since the breakpoint definitely triggers.

Reply:
It's probably not a codecov problem. I downloaded the jacoco report from [this GitHub Actions run](https://github.com/JanusGraph/janusgraph/pull/2516/checks?check_run_id=2790668124) and loaded into Intellij. It shows that JanusGraphPropertiesStep has 0% coverage.
Can you share which JanusGraph test class would lead to[ this relevant test](https://github.com/apache/tinkerpop/blob/6bb5c19e5b3845a1547f079151017861b47e0828/gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/TraversalInterruptionTest.java#L121)?

Reply:
I used [this one](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-inmemory/src/test/java/org/janusgraph/blueprints/inmemory/process/InMemoryMultiQueryJanusGraphProcessTest.java) for debugging. It executes a number of TinkerPop Tests, including the one you mentioned. Executing this test class leads to the breakpoint being triggered.

Reply:
Thanks, indeed. I also verified that test indeed was executed in CI. No idea why jacoco does not include `JanusGraphPropertiesStep` in report...